File: articles/calculus-for-systems-modeling/articles/definite-integrals-and-total-change/julia/definite_integral_audit.jl
net_rate(t) = 4.0 * sin(t / 2.0) + 1.0

function trapezoid_integral(values, times)
    total = 0.0
    for i in 1:(length(times)-1)
        dt = times[i+1] - times[i]
        if dt <= 0
            error("Times must be strictly increasing.")
        end
        total += 0.5 * (values[i] + values[i+1]) * dt
    end
    return total
end

times = collect(0.0:0.5:4.0)
rates = [net_rate(t) for t in times]
signed_accumulation = trapezoid_integral(rates, times)
absolute_accumulation = trapezoid_integral(abs.(rates), times)

println("interval_start,interval_end,method,signed_accumulation,absolute_accumulation,warning")
println("$(first(times)),$(last(times)),trapezoidal approximation,$signed_accumulation,$absolute_accumulation,signed and absolute accumulation separated")

File: articles/calculus-for-systems-modeling/articles/definite-integrals-and-total-change/r/definite_integral_audit.R
net_rate <- function(t) {
  4 * sin(t / 2) + 1
}

trapezoid_integral <- function(values, times) {
  total <- 0
  for (i in seq_len(length(times) - 1)) {
    dt <- times[i + 1] - times[i]
    if (dt <= 0) {
      stop("Times must be strictly increasing.")
    }
    total <- total + 0.5 * (values[i] + values[i + 1]) * dt
  }
  total
}

times <- seq(0, 4, by = 0.5)
rates <- net_rate(times)

signed_accumulation <- trapezoid_integral(rates, times)
absolute_accumulation <- trapezoid_integral(abs(rates), times)

warning <- ""
if (any(rates < 0) && abs(signed_accumulation) < absolute_accumulation) {
  warning <- "signed accumulation includes cancellation"
}

result <- data.frame(
  interval_start = min(times),
  interval_end = max(times),
  method = "trapezoidal approximation",
  signed_accumulation = signed_accumulation,
  absolute_accumulation = absolute_accumulation,
  unit_check = "rate units times time units = accumulated quantity units",
  interpretation = "signed accumulation estimates net change; absolute accumulation estimates total activity",
  warning = warning
)

dir.create("outputs/tables", recursive = TRUE, showWarnings = FALSE)
write.csv(result, "outputs/tables/r_definite_integral_audit.csv", row.names = FALSE)
print(result)
